sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/ownership_manager.py

```diff
@@ -157,33 +138,60 @@ def _balance_ownership(self, ownership_list, all_partition_ids):
         # owners_count is the number of active owners. If self.owner_id is not yet among the active owners,
         # then plus 1 to include self. This will make owners_count >= 1.
         owners_count = len(active_ownership_by_owner) + (
             0 if self.owner_id in active_ownership_by_owner else 1
         )
         expected_count_per_owner = all_partition_count // owners_count
+        # Py2 math.ceil() returns float, a/b return int if not divisable.
+        # Py3 math.ceil() returns int, a/b return float if not divisable.
+        max_count_per_owner = int(math.ceil(all_partition_count*1.0 / owners_count))
         # end of calculating expected count per owner
 
         to_claim = active_ownership_self
-        if len(active_ownership_self) < expected_count_per_owner:
-            # Either claims an inactive partition, or steals from other owners
-            if claimable_partition_ids:  # claim an inactive partition if there is
-                random_partition_id = random.choice(claimable_partition_ids)
-                random_chosen_to_claim = ownership_dict.get(
-                    random_partition_id,
-                    {
-                        "fully_qualified_namespace": self.fully_qualified_namespace,
-                        "partition_id": random_partition_id,
-                        "eventhub_name": self.eventhub_name,
-                        "consumer_group": self.consumer_group,
-                    },
+        if len(active_ownership_self) < max_count_per_owner:
+            to_try_steal = True
+            if self.load_balancing_strategy is LoadBalancingStrategy.GREEDY:
+                # Greedily claim more partitions if there are claimable partitions
+                to_greedy_claim_ids = random.sample(
+                    claimable_partition_ids, k=min(
+                        max_count_per_owner - len(active_ownership_self), len(claimable_partition_ids)
+                    )
                 )
-                random_chosen_to_claim["owner_id"] = self.owner_id
-                to_claim.append(random_chosen_to_claim)
-            else:  # steal from another owner that has the most count
+                if to_greedy_claim_ids:
+                    for pid in to_greedy_claim_ids:
+                        random_chosen_to_claim = ownership_dict.get(
+                            pid,
+                            {
+                                "fully_qualified_namespace": self.fully_qualified_namespace,
+                                "partition_id": pid,
+                                "eventhub_name": self.eventhub_name,
+                                "consumer_group": self.consumer_group,
+                            },
+                        )
+                        random_chosen_to_claim["owner_id"] = self.owner_id
+                        to_claim.append(random_chosen_to_claim)
+                    to_try_steal = False  # already greedily got at least one
```

Comment:
Let's say we greedily claimed fewer messages than our expected_count_per_owner, we still don't want to try and steal?  (I would think this would help it converge faster, as long as you prevent a churning situation where you keep stealing one element from each other back and forth, and that scenario would be a risk across the board and not just in this flow.  If that's already dealt with and I just haven't read far enough feel free to ignore this)

Reply:
From our OOB discussion; we may be able to be more eager in stealing and will consider this; and churning shouldn't be an issue if the expected vs. max bounding plays out properly, since you if e.g. 10 partitions and 9 consumers, expected would be 1 meaning you don't pass that spare around in a cycle; and at the inflection point of 5 then there's no reason to steal since expected would be 2 for the others.